src/groups/mqb/mqbblp/mqbblp_relayqueueengine.cpp

```diff
@@ -1682,17 +1676,472 @@ bool RelayQueueEngine::subscriptionId2upstreamSubQueueId(
+                                                        &d_options_sp,
+                                                        &d_attributes,
+                                                        d_iterator->first);
+        BSLS_ASSERT_SAFE(mqbi::StorageResult::e_SUCCESS == rc);
+        static_cast<void>(rc);  // suppress compiler warning
+        return true;            // RETURN
+    }
+    return false;
+}
+
+// CREATORS
+RelayQueueEngine_PushStorageIterator::RelayQueueEngine_PushStorageIterator(
+    mqbi::Storage*              storage,
+    PushStream*                 owner,
+    const PushStream::iterator& initialPosition)
+: d_storage_p(storage)
+, d_iterator(initialPosition)
+, d_attributes()
+, d_appData_sp()
+, d_options_sp()
+, d_owner_p(owner)
+, d_currentElement(0)
+, d_currentOrdinal(mqbi::Storage::k_INVALID_ORDINAL)
+{
+    BSLS_ASSERT_SAFE(d_storage_p);
+    BSLS_ASSERT_SAFE(d_owner_p);
+}
+
+RelayQueueEngine_PushStorageIterator::~RelayQueueEngine_PushStorageIterator()
+{
+    // NOTHING
+}
+
+unsigned int RelayQueueEngine_PushStorageIterator::numApps() const
+{
+    BSLS_ASSERT_SAFE(!atEnd());
+    return d_iterator->second.numElements();
+}
+
+void RelayQueueEngine_PushStorageIterator::removeCurrentElement()
+{
+    BSLS_ASSERT_SAFE(!atEnd());
+    BSLS_ASSERT_SAFE(d_currentElement);
+
+    PushStream::Element* del = d_currentElement;
+
+    // still keep the same ordinal numbering
+    d_currentElement = d_currentElement->next();
+    ++d_currentOrdinal;
+
+    d_owner_p->remove(del);
+    d_owner_p->destroy(del, true);
+    // doKeepGuid because of the d_iterator
+
+    if (d_iterator->second.numElements() == 0) {
+        // d_currentElement->eraseFromStream(d_owner_p->d_stream);
+
+        BSLS_ASSERT_SAFE(d_currentElement == 0);
+    }
+}
+
+// MANIPULATORS
+bool RelayQueueEngine_PushStorageIterator::advance()
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(!atEnd());
+
+    clear();
+
+    if (d_iterator->second.numElements() == 0) {
+        d_iterator = d_owner_p->d_stream.erase(d_iterator);
+    }
+    else {
+        ++d_iterator;
+    }
+
+    d_currentOrdinal = mqbi::Storage::k_INVALID_ORDINAL;
+
+    return !atEnd();
+}
+
+void RelayQueueEngine_PushStorageIterator::reset(
+    const bmqt::MessageGUID& where)
+{
+    clear();
+
+    if (where.isUnset()) {
+        // Reset iterator to beginning
+        d_iterator = d_owner_p->d_stream.begin();
+    }
+    else {
+        d_iterator = d_owner_p->d_stream.find(where);
+    }
+
+    d_currentOrdinal = mqbi::Storage::k_INVALID_ORDINAL;
+}
+
+// ACCESSORS
+const bmqt::MessageGUID& RelayQueueEngine_PushStorageIterator::guid() const
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(!atEnd());
+
+    return d_iterator->first;
+}
+
+PushStream::Element*
+RelayQueueEngine_PushStorageIterator::element(unsigned int appOrdinal) const
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(!atEnd());
+
+    if (d_currentOrdinal > appOrdinal) {
```

Comment:
Is this logic okay if `appOrdinal` is the invalid ordinal?  It seems like that will end up iterating `d_currentOrdinal` potentially until we hit an `ASSERT(d_currentElement)`.  Is this what we want, and if so, does it make sense to add a precondition/`ASSERT` that `appOrdinal` is not the invalid ordinal?

Reply:
This will assert if `appOrdinal` is not valid (not just `k_INVALID_ORDINAL`) at  ASSERT(d_currentElement)
Adding assert does not hurt